i#731,i#3271: Keep IR cti+copied bits valid and re-relativized

derekbruening opened the pull request:
Adds rip-relative information tracking at IR levels 1-3 for
instruction references on x86, extending the existing tracking for
data references.

Adds decode_sizeof_ex() and instr_get_rel_data_or_instr_target() to
support rip-relative displacement and target handling.

Changes decode_from_copy() to preserve raw bits for all instructions,
eliminating problems where untracked encoding features are lost such
as in #4017.

Documents the changes and describes how to avoid the new behavior.
Does just that for intra-sequence cti's in the rseq native code copy.

Adds test cases to api.ir.

Implementing the same thing for AArchXX is left unimplemented, tracked
by i#4016.

Issue: #731, #3271, #3339, #4016, #4017
Fixes #731
Fixes #3271
Fixes #4017

derekbruening commented:
(Making sure the CI is green before requesting review since this has potential to break things.)

derekbruening commented:
All issues resolved.  PTAL.  See https://github.com/DynamoRIO/dynamorio/issues/4017#issuecomment-574782637 for logic behind tackling the full #731.

Carrotman42 commented on core/arch/x86/decode.c:
```diff
@@ -2654,27 +2657,19 @@ decode_common(dcontext_t *dcontext, byte *pc, byte *orig_pc, instr_t *instr)
+    IF_X64(ASSERT(CHECK_TRUNCATE_TYPE_uint(next_pc - pc)));
+    instr_set_raw_bits(instr, pc, (uint)(next_pc - pc));
+    if (di.disp_abs > di.start_pc) {
+        CLIENT_ASSERT(di.disp_abs > di.start_pc, "decode: internal rip-rel error");
```
The expression inside the 'if' is the same as this assert, so the assert will never be false afaict. Either take out the assert, or perhaps you meant for the 'if' to be different?

derekbruening replied:
Removed.

Carrotman42 commented on core/arch/x86/encode.c:
```diff
@@ -2722,14 +2717,20 @@ copy_and_re_relativize_raw_instr(dcontext_t *dcontext, instr_t *instr, byte *dst
+            *((char *)dst_pc) = (char)new_offs;
+        } else {
+            ASSERT(rip_rel_pos + 4 <= instr->length);
+            *((int *)dst_pc) = (int)new_offs;
```
Should we assert that this truncates to a singed int? I imagine 'int' is not always pointer sized, but I don't know for sure.

derekbruening replied:
Assert added.  int is not pointer-sized for 64-bit gcc/clang/cl.

Carrotman42 commented on suite/tests/api/ir_x86.c:
```diff
@@ -2168,6 +2168,97 @@ test_reg_exact_reads(void *dc)
+    byte buf_dec_enc[] = { opc1, opc2,
+                           /* disp32 of 0 which targets the next PC. */
+                           0x00, 0x00, 0x00, 0x00,
+                           /* We encode here. */ 0x90, 0x90, 0x90, 0x90, 0x90, 0x90 };
```
I'm guessing this line-joining is due to clang-format? Consider adding whitespace after the "We encode here." comment so that clang-format decides to wrap like the previous line is wrapped.

(alternatively we could use a line-comment, but that's against the existing style)

derekbruening replied:
Newline successfully added and preserved by clang-format.

derekbruening commented:
PTAL.  Sorry, I pre-emptively merged in upstream to deal w/ a conflict, so the "changes since last review" button may not be a good view: but selecting just the final commit from the commit selection dropdown on the diff page will show the review changes.
